### ISSUE TITLE
.Select Eliminate where clauses effects

### ISSUE THREAD

**Issue by imansafari1991** [redis/redis-om-dotnet#307]
In my search scenario, I have a bunch of arbitrary filters that I implement chain of where clauses as follows(of course I try to reconstruct my issue by customer model in the source):

`
            customers.Insert(new Customer{
                FirstName = "James",
                LastName = "Bond",
                Email = "bondjamesbond@email.com",
                Age = 68
            });
            customers.Insert(new Customer
            {
                FirstName = "Iman",
                LastName = "Safari",
                Email = "imansafari33@gmail.com",
                Age =32
            });
            customers.Insert(new Customer
            {
                FirstName = "Tom",
                LastName = "Cruise",
                Email = "TomCruise@email.com",
                Age = 61
            });`

I wrote my filter as follows:
`    int? minAge = 60;

            int? maxAge = 63;
            if (minAge != null)
            {
                customers = customers.Where(p => p.Age >= minAge);
            }

            if (maxAge != null)
            {
                customers = customers.Where(p => p.Age <= maxAge);
            }`

when I used ToList() I got the expected to answer that is "tom cruise":
` var customerList = customers.ToList();`

When I want to select specific property like "firstname" , the where clause effect is disappeared and returns three objects!!

`var customerNames=customers.Select(p=>p.FirstName).ToList();`

Is there anything wrong with me?

Thank you

**Comment by zulander1** [comment]
You have the full code? It shouldn't do that!

**Comment by imansafari1991** [comment]
> You have the full code? It shouldn't do that!
Hi, here you are.
` public class Program
    {
        [Document(StorageType = StorageType.Json)]
        public class Customer
        {
            [Indexed] public string FirstName { get; set; }
            [Indexed] public string LastName { get; set; }
            [Indexed] public string Email { get; set; }
            [Indexed(Aggregatable = true)] public int Age { get; set; }
            [Indexed(Aggregatable = true)] public GeoLoc Home { get; set; }
        }
        static void Main(string[] args)
        {
            // connect
            var provider = new RedisConnectionProvider("redis://localhost:6379");
            var connection = provider.Connection;
            var customers = provider.RedisCollection<Customer>();
            
            
            // Create index
            connection.CreateIndex(typeof(Customer));

            // Get Index info
            var indexinfo = connection.GetIndexInfo(typeof(Customer));

            // Insert Object
            customers.Insert(new Customer{
                FirstName = "James",
                LastName = "Bond",
                Email = "bondjamesbond@email.com",
                Age = 68
            });
            customers.Insert(new Customer
            {
                FirstName = "Iman",
                LastName = "Safari",
                Email = "imansafari33@gmail.com",
                Age =32
            });
            customers.Insert(new Customer
            {
                FirstName = "Tom",
                LastName = "Cruise",
                Email = "TomCruise@email.com",
                Age = 61
            });


            int? minAge = 60;

            int? maxAge = 63;
       
            var ages = new int[] { 61, 32 };
       
            var customerNames=customers.Select(p=>p.Age).ToList();
        }
    }`

**Comment by slorello89** [comment]
I see it - https://github.com/redis/redis-om-dotnet/blob/main/src/Redis.OM/SearchExtensions.cs#L119 - it's passing null rather than propagating forward the filter, I'll have to think on this, might be a non-trivial fix because the `Select` mutates the return type from `IRedisCollection<T>` to `IRedisCollection<TR>` - those boolean expressions are based on `T` lol.

**Comment by AlexanderAnatolich** [comment]
I had a close issue like yours.
I resolve it like that:

Building query
```
private Expression<Func<T, bool>> BuildSearchExpression(U model)
        {
            Expression<Func<T, bool>> expression = default;
            if (model.Field.HasValue)
            {
                JoinAndExpression(ref expression, (it) => it.Field> model.Field.Value);
            }

            if (model.Field2.HasValue)
            {
                JoinAndExpression(ref expression, (it) => it.Field2== model.Field2.Value);
            }            
            return expression;
        }
```
```
public static void JoinAndExpression(
            ref Expression<Func<T, bool>> root,
            Expression<Func<T, bool>> nextExpression)
        {
            if (nextExpression == null)
            {
                return;
            }

            if (root == null)
            {
                root = nextExpression;
                return;
            }

            root = root.And(nextExpression);
        }
```

Query execution:

```
var resulted = redisCollection;

                if (filters != null)
                {
                    resulted = redisCollection.Where(filters);
                }

                if (skip.HasValue)
                {
                    resulted = resulted.Skip(skip.Value);
                }

                if (count.HasValue)
                {
                    resulted = resulted.Take(count.Value);
                }
var result  = resulted.ToList()
```

P.S. `root.And(nextExpression)` `And` is from `LinqKit`

**Comment by slorello89** [comment]
Yeah - looks like I broke this when I implemented #157 over the summer (basically we lose the context of the primary type when building the big boolean expression down the pipe when we move over to a SELECT)
